In [3]:
data=datasets::mtcars

In [4]:
head(data,10)

mpg  cyl disp  hp  drat wt    qsec  vs am gear carb
Mazda RX4         21.0 6   160.0 110 3.90 2.620 16.46 0  1  4    4   
Mazda RX4 Wag     21.0 6   160.0 110 3.90 2.875 17.02 0  1  4    4   
Datsun 710        22.8 4   108.0  93 3.85 2.320 18.61 1  1  4    1   
Hornet 4 Drive    21.4 6   258.0 110 3.08 3.215 19.44 1  0  3    1   
Hornet Sportabout 18.7 8   360.0 175 3.15 3.440 17.02 0  0  3    2   
Valiant           18.1 6   225.0 105 2.76 3.460 20.22 1  0  3    1   
Duster 360        14.3 8   360.0 245 3.21 3.570 15.84 0  0  3    4   
Merc 240D         24.4 4   146.7  62 3.69 3.190 20.00 1  0  4    2   
Merc 230          22.8 4   140.8  95 3.92 3.150 22.90 1  0  4    2   
Merc 280          19.2 6   167.6 123 3.92 3.440 18.30 1  0  4    4

In [5]:
dim(data)

[1] 32 11

In [7]:
summary(data)

      mpg             cyl             disp             hp       
 Min.   :10.40   Min.   :4.000   Min.   : 71.1   Min.   : 52.0  
 1st Qu.:15.43   1st Qu.:4.000   1st Qu.:120.8   1st Qu.: 96.5  
 Median :19.20   Median :6.000   Median :196.3   Median :123.0  
 Mean   :20.09   Mean   :6.188   Mean   :230.7   Mean   :146.7  
 3rd Qu.:22.80   3rd Qu.:8.000   3rd Qu.:326.0   3rd Qu.:180.0  
 Max.   :33.90   Max.   :8.000   Max.   :472.0   Max.   :335.0  
      drat             wt             qsec             vs        
 Min.   :2.760   Min.   :1.513   Min.   :14.50   Min.   :0.0000  
 1st Qu.:3.080   1st Qu.:2.581   1st Qu.:16.89   1st Qu.:0.0000  
 Median :3.695   Median :3.325   Median :17.71   Median :0.0000  
 Mean   :3.597   Mean   :3.217   Mean   :17.85   Mean   :0.4375  
 3rd Qu.:3.920   3rd Qu.:3.610   3rd Qu.:18.90   3rd Qu.:1.0000  
 Max.   :4.930   Max.   :5.424   Max.   :22.90   Max.   :1.0000  
       am              gear            carb      
 Min.   :0.0000   Min.   :3.000  

One way ANOVA
Since we only have one dataset, we can use it to create multiple samples to perform both methods of ANOVA
We will create 3 datasets of 6 rows each and then perform our tests
For One Way ANOVA we will use qsec column of the dataframe, assuming α=0.05
NULL HYPOTHESIS=> All 3 samples belong to the same population, ie, mu1=mu2=mu3
ALTERANTE HYPOTHESIS=> Not all samples belong to the same population, or, at least 2 of the samples have diff means

In [8]:
set.seed(5)
sample1 = data[sample(nrow(data),6),]
sample1


mpg  cyl disp  hp  drat wt    qsec  vs am gear carb
Mazda RX4 Wag      21.0 6   160.0 110 3.90 2.875 17.02 0  1  4    4   
Merc 280C          17.8 6   167.6 123 3.92 3.440 18.90 1  0  4    4   
Pontiac Firebird   19.2 8   400.0 175 3.08 3.845 17.05 0  0  3    2   
Cadillac Fleetwood 10.4 8   472.0 205 2.93 5.250 17.98 0  0  3    4   
Maserati Bora      15.0 8   301.0 335 3.54 3.570 14.60 0  1  5    8   
Ferrari Dino       19.7 6   145.0 175 3.62 2.770 15.50 0  1  5    6

In [9]:
set.seed(8)
sample2 = data[sample(nrow(data),6),]
sample2


mpg  cyl disp  hp  drat wt    qsec  vs am gear carb
Volvo 142E         21.4 4   121.0 109 4.11 2.780 18.60 1  1  4    2   
Toyota Corolla     33.9 4    71.1  65 4.22 1.835 19.90 1  1  4    1   
AMC Javelin        15.2 8   304.0 150 3.15 3.435 17.30 0  0  3    2   
Mazda RX4 Wag      21.0 6   160.0 110 3.90 2.875 17.02 0  1  4    4   
Cadillac Fleetwood 10.4 8   472.0 205 2.93 5.250 17.98 0  0  3    4   
Merc 450SE         16.4 8   275.8 180 3.07 4.070 17.40 0  0  3    3

In [10]:
set.seed(12)
sample3 = data[sample(nrow(data),6),]
sample3

mpg  cyl disp  hp  drat wt    qsec  vs am gear carb
Mazda RX4 Wag       21.0 6   160.0 110 3.90 2.875 17.02 0  1  4    4   
Fiat X1-9           27.3 4    79.0  66 4.08 1.935 18.90 1  1  4    1   
Lincoln Continental 10.4 8   460.0 215 3.00 5.424 17.82 0  0  3    4   
AMC Javelin         15.2 8   304.0 150 3.15 3.435 17.30 0  0  3    2   
Porsche 914-2       26.0 4   120.3  91 4.43 2.140 16.70 0  1  5    2   
Merc 450SLC         15.2 8   275.8 180 3.07 3.780 18.00 0  0  3    3

In [11]:
sample1$group<-"sample1"
sample1
sample2$group<-"sample2"
sample2
sample3$group<-"sample3"
sample3

mpg  cyl disp  hp  drat wt    qsec  vs am gear carb group  
Mazda RX4 Wag      21.0 6   160.0 110 3.90 2.875 17.02 0  1  4    4    sample1
Merc 280C          17.8 6   167.6 123 3.92 3.440 18.90 1  0  4    4    sample1
Pontiac Firebird   19.2 8   400.0 175 3.08 3.845 17.05 0  0  3    2    sample1
Cadillac Fleetwood 10.4 8   472.0 205 2.93 5.250 17.98 0  0  3    4    sample1
Maserati Bora      15.0 8   301.0 335 3.54 3.570 14.60 0  1  5    8    sample1
Ferrari Dino       19.7 6   145.0 175 3.62 2.770 15.50 0  1  5    6    sample1

mpg  cyl disp  hp  drat wt    qsec  vs am gear carb group  
Volvo 142E         21.4 4   121.0 109 4.11 2.780 18.60 1  1  4    2    sample2
Toyota Corolla     33.9 4    71.1  65 4.22 1.835 19.90 1  1  4    1    sample2
AMC Javelin        15.2 8   304.0 150 3.15 3.435 17.30 0  0  3    2    sample2
Mazda RX4 Wag      21.0 6   160.0 110 3.90 2.875 17.02 0  1  4    4    sample2
Cadillac Fleetwood 10.4 8   472.0 205 2.93 5.250 17.98 0  0  3    4    sample2
Merc 450SE         16.4 8   275.8 180 3.07 4.070 17.40 0  0  3    3    sample2

mpg  cyl disp  hp  drat wt    qsec  vs am gear carb group  
Mazda RX4 Wag       21.0 6   160.0 110 3.90 2.875 17.02 0  1  4    4    sample3
Fiat X1-9           27.3 4    79.0  66 4.08 1.935 18.90 1  1  4    1    sample3
Lincoln Continental 10.4 8   460.0 215 3.00 5.424 17.82 0  0  3    4    sample3
AMC Javelin         15.2 8   304.0 150 3.15 3.435 17.30 0  0  3    2    sample3
Porsche 914-2       26.0 4   120.3  91 4.43 2.140 16.70 0  1  5    2    sample3
Merc 450SLC         15.2 8   275.8 180 3.07 3.780 18.00 0  0  3    3    sample3

In [12]:
sample<-rbind(sample1,sample2,sample3)
sample


mpg  cyl disp  hp  drat wt    qsec  vs am gear carb group  
Mazda RX4 Wag       21.0 6   160.0 110 3.90 2.875 17.02 0  1  4    4    sample1
Merc 280C           17.8 6   167.6 123 3.92 3.440 18.90 1  0  4    4    sample1
Pontiac Firebird    19.2 8   400.0 175 3.08 3.845 17.05 0  0  3    2    sample1
Cadillac Fleetwood  10.4 8   472.0 205 2.93 5.250 17.98 0  0  3    4    sample1
Maserati Bora       15.0 8   301.0 335 3.54 3.570 14.60 0  1  5    8    sample1
Ferrari Dino        19.7 6   145.0 175 3.62 2.770 15.50 0  1  5    6    sample1
Volvo 142E          21.4 4   121.0 109 4.11 2.780 18.60 1  1  4    2    sample2
Toyota Corolla      33.9 4    71.1  65 4.22 1.835 19.90 1  1  4    1    sample2
AMC Javelin         15.2 8   304.0 150 3.15 3.435 17.30 0  0  3    2    sample2
Mazda RX4 Wag1      21.0 6   160.0 110 3.90 2.875 17.02 0  1  4    4    sample2
Cadillac Fleetwood1 10.4 8   472.0 205 2.93 5.250 17.98 0  0  3    4    sample2
Merc 450SE          16.4 8   275.8 180 3.07 4.070 17.40 0  0  3    3    sample2
Mazda RX4 Wag2      21.0 6   160.0 110 3.90 2.875 17.02 0  1  4    4    sample3
Fiat X1-9           27.3 4    79.0  66 4.08 1.935 18.90 1  1  4    1    sample3
Lincoln Continental 10.4 8   460.0 215 3.00 5.424 17.82 0  0  3    4    sample3
AMC Javelin1        15.2 8   304.0 150 3.15 3.435 17.30 0  0  3    2    sample3
Porsche 914-2       26.0 4   120.3  91 4.43 2.140 16.70 0  1  5    2    sample3
Merc 450SLC         15.2 8   275.8 180 3.07 3.780 18.00 0  0  3    3    sample3

In [13]:
res.aov<-aov(qsec~group,data=sample)
summary(res.aov)


            Df Sum Sq Mean Sq F value Pr(>F)
group        2  4.398   2.199   1.546  0.245
Residuals   15 21.339   1.423               

Since P value(0.245)>alpha(0.05)
Therfore,We accept the null hypothesis that all three samples belong to same polulation

In [14]:
install.packages("plyr", dependencies=T)
install.packages("dplyr")

Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)

also installing the dependencies ‘lazyeval’, ‘rex’, ‘abind’, ‘covr’, ‘doParallel’, ‘foreach’, ‘iterators’, ‘itertools’


Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)



In [15]:
library('plyr')
library('dplyr')


Attaching package: ‘dplyr’


The following objects are masked from ‘package:plyr’:

    arrange, count, desc, failwith, id, mutate, rename, summarise,
    summarize


The following objects are masked from ‘package:stats’:

    filter, lag


The following objects are masked from ‘package:base’:

    intersect, setdiff, setequal, union




In [17]:
anova<-function(data,col,group){
 stats<-aggregate(col, list(group), FUN=mean)
 means<-stats[,2]
 cnt<-count(data,group)[,2]
 k<-nrow(stats)
 mean_total=sum(col)/nrow(data)
 dof1=k-1
 dof2=nrow(data)-k
 bss<-0
 for(i in 1:k){
 bss=bss+cnt[i]*(means[i]-mean_total)**2
 }
 msb=bss/dof1
 wss<-0
 for(i in 1:nrow(data)){
 curr_mean=means[stats[1]==group[i]]
 wss=wss+(col[i]-curr_mean)**2
 }
 msw=wss/dof2
 fcal=msb/msw
 return(fcal)
}


In [18]:
anova(sample,sample$qsec,sample$group)

[1] 1.545915

1.54591494423629
Same F-value as the inbuilt one way ANOVA function(our user defined function works well)

Two way ANOVA
We will use the same sample dataframe that we built earlier using random values
This time we will be using column-disp against am and group columns
NULL HYPOTHESIS A=>Mean of disp is same for all groups
ALTERANTE HYPOTHESIS A=>At least two of the groups have different means
NULL HYPOTHESIS B=>Mean of disp is same for both values of am
ALTERANTE HYPOTHESIS=>Both values of am have different means

In [19]:
aov2<-aov(disp~group+am,data=sample)

In [20]:
summary(aov2)

            Df Sum Sq Mean Sq F value   Pr(>F)    
group        2   6622    3311   0.379 0.691160    
am           1 182771  182771  20.937 0.000432 ***
Residuals   14 122217    8730                     
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

From first p value(0.69>0.05) We can conclude that the mean for all three groups is same
From second p value(0.0004<0.05) We can reject null hypothesis B, so both values of am have different means

In [21]:
two_way_anova<-function(data,col,group1,group2){
 statsy<-aggregate(col,list(group1),FUN=mean)
 statsx<-aggregate(col,list(group2),FUN=mean)
 meansy<-statsy[,2]
 meansx<-statsx[,2]
 cnty<-aggregate(col,list(group1),FUN=length)[,2]
 cntx<-aggregate(col,list(group2),FUN=length)[,2]
 grand_mean=sum(col)/nrow(data)
 p=nrow(statsy)
 q=nrow(statsx)
 ssa=q*(sum((meansy-grand_mean)**2))*3
 ssb=p*(sum((meansx-grand_mean)**2))*3
 sse<-0
 for(i in 1:nrow(data)){
 curr_meany=meansy[statsy[1]==group1[i]]
 curr_meanx=meansx[statsx[1]==group2[i]]
 sse=sse+((col[i]-curr_meany-curr_meanx+grand_mean)**2)
 }
 msa=ssa/(p-1)
 msb=ssb/(q-1)
 mse=sse/(cnty[1]+cntx[1]-1)
 f1=msa/mse
 f2=msb/mse
 return(list(f1,f2))
}


In [22]:
two_way_anova(sample,sample$disp,sample$group,sample$am)


[[1]]
[1] 0.3793037

[[2]]
[1] 20.93651

We get the same f values as we saw earlier,Hence the same conclusion